File: objc/action/SFActionInstant.m
//
//  SFActionInstant.m
//  SpriteForest2
//
//  Created by Moky on 12-10-1.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "SFActionInstant.h"

#import "coconut2d.h"


@implementation SFActionInstant

- (CCAction *) ccAction
{
	CNLog(@"override me");
	return nil;
//	return [CCActionInstant action];
}

@end


@implementation SFShow

- (CCAction *) ccAction
{
	return [CCShow action];
}

@end


@implementation SFHide

- (CCAction *) ccAction
{
	return [CCHide action];
}

@end


@implementation SFToggleVisibility

- (CCAction *) ccAction
{
	return [CCToggleVisibility action];
}

@end


@implementation SFFlipX

- (CCAction *) ccAction
{
	return [CCFlipX actionWithFlipX:YES];
}

@end


@implementation SFFlipY

- (CCAction *) ccAction
{
	return [CCFlipY actionWithFlipY:YES];
}

@end


@implementation SFPlace

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		position_ = [dict objectForKey:@"Position"] ? CCPointFromString([dict objectForKey:@"Position"]) : CGPointZero;
	}
	return self;
}

- (CCAction *) ccAction
{
	return [CCPlace actionWithPosition:position_];
}

@end


#pragma mark - Action Call Function


@implementation SFCallFunc

@synthesize targetCallback = targetCallback_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		self.targetCallback = [dict objectForKey:@"Target"];
		
		NSString * selector = [dict objectForKey:@"Selector"];
		if ([selector isKindOfClass:[NSString class]])
		{
			selector_ = NSSelectorFromString(selector);
		}
		else
		{
			CNLog(@"ERROR: invalid selector, dict = %@", dict);
		}
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.targetCallback = nil;
		selector_ = nil;
	}
	return self;
}

- (void) dealloc
{
	[targetCallback_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (!selector_ || ![targetCallback_ respondsToSelector:selector_])
	{
		CNLog(@"Error: invalid selector for this target.");
		return [super ccAction];
	}
	return [CCCallFunc actionWithTarget:targetCallback_ selector:selector_];
}

@end


//@implementation SFCallFuncN
//
//<#methods#>
//
//@end
//
//
//@implementation SFCallFuncND
//
//<#methods#>
//
//@end


@implementation SFCallFuncO

@synthesize object = object_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		self.object = [dict objectForKey:@"Object"];
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.object = nil;
	}
	return self;
}

- (void) dealloc
{
	[object_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (!selector_ || ![targetCallback_ respondsToSelector:selector_])
	{
		CNLog(@"Error: invalid selector for this target: %@", targetCallback_);
		return [super ccAction];
	}
	return [CCCallFuncO actionWithTarget:targetCallback_ selector:selector_ object:object_];
}

@end


//@implementation SFCallBlock
//
//<#methods#>
//
//@end
//
//
//@implementation SFCallBlockN
//
//<#methods#>
//
//@end

File: spriteforest-x/action/SFActionInstant.cpp
//
//  SFActionInstant.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-11-27.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "SFActionInstant.h"

NS_SF_BEGIN

USING_NS_CC;
USING_NS_CN;

#pragma mark SFActionInstant

CCAction * SFActionInstant::ccAction(void)
{
	CNLog("override me!");
	return NULL;
}

#pragma mark SFShow

CCAction * SFShow::ccAction(void)
{
	return CCShow::action();
}

#pragma mark SFHide

CCAction * SFHide::ccAction(void)
{
	return CCHide::action();
}

#pragma mark SFToggleVisibility

CCAction * SFToggleVisibility::ccAction(void)
{
	return CCToggleVisibility::action();
}

#pragma mark SFFlipX

CCAction * SFFlipX::ccAction(void)
{
	return CCFlipX::action();
}

#pragma mark SFFlipY

CCAction * SFFlipY::ccAction(void)
{
	return CCFlipY::action();
}

#pragma mark SFPlace

bool SFPlace::initWithDictionary(CCDictionary * pDict)
{
	if (!SFActionInstant::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	CCString * pPosition = (CCString *)pDict->objectForKey("Position");
	if (pPosition)
	{
		m_tPosition = CCPointFromString(pPosition->getCString());
	}
	return true;
}

CCAction * SFPlace::ccAction(void)
{
	return CCPlace::actionWithPosition(m_tPosition);
}

#pragma mark - Action call back

#pragma mark SFCallFunc

SFCallFunc::SFCallFunc(void)
: SFActionInstant()
, m_pTargetCallback(NULL)
, m_pSelector(NULL)
{
	
}

SFCallFunc::~SFCallFunc(void)
{
	setTargetCallback(NULL);
}

bool SFCallFunc::init(void)
{
	setTargetCallback(NULL);
	setSelector(NULL);
	return true;
}

bool SFCallFunc::initWithDictionary(CCDictionary * pDict)
{
	if (!SFActionInstant::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	CCObject * pTarget = pDict->objectForKey("Target");
	if (!pTarget)
	{
		pTarget = pDict->objectForKey("ss_Target");
	}
	setTargetCallback(pTarget);
	setSelector((CCString *)pDict->objectForKey("Selector"));
	return true;
}

CCAction * SFCallFunc::ccAction(void)
{
	if (!m_pTargetCallback || !m_pSelector)
	{
		CNLog("error");
		return NULL;
	}
	CNSelectorDelegate * pCallback = dynamic_cast<CNSelectorDelegate *>(m_pTargetCallback);
	if (!pCallback)
	{
		CNLog("cannot get selector: %s", m_pSelector->getCString());
		return NULL;
	}
	SEL_CallFunc selector = pCallback->getSelector(m_pSelector->getCString());
	if (!selector)
	{
		CNLog("no such selector: %s", m_pSelector->getCString());
		return NULL;
	}
	CNLog("selector: %s", m_pSelector->getCString());
	return CCCallFunc::actionWithTarget(m_pTargetCallback, selector);
}

#pragma mark SFCallFuncO

SFCallFuncO::SFCallFuncO(void)
: SFCallFunc()
, m_pObject(NULL)
{
	
}

SFCallFuncO::~SFCallFuncO(void)
{
	setObject(NULL);
}

bool SFCallFuncO::init(void)
{
	setObject(NULL);
	return true;
}

bool SFCallFuncO::initWithDictionary(CCDictionary * pDict)
{
	if (!SFCallFunc::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	setObject(pDict->objectForKey("Object"));
	return true;
}

CCAction * SFCallFuncO::ccAction(void)
{
	CCObject * pTargetCallback = getTargetCallback();
	CCString * pSelector = getSelector();
	if (!pTargetCallback || !pSelector)
	{
		CNLog("error");
		return NULL;
	}
	CNSelectorDelegate * pCallback = dynamic_cast<CNSelectorDelegate *>(pTargetCallback);
	if (!pCallback)
	{
		CNLog("cannot get selector: %s", pSelector->getCString());
		return NULL;
	}
	SEL_CallFuncO selector = pCallback->getSelectorO(pSelector->getCString());
	if (!selector)
	{
		CNLog("no such selector: %s", pSelector->getCString());
		return NULL;
	}
	return CCCallFuncO::actionWithTarget(pTargetCallback, selector, m_pObject);
}

NS_SF_END
